### PR TITLE
scenario.utils.capture_events

### PR DESCRIPTION
**Pull request by PietroPasotti**
This seems like a common-enough use case that we should offer it in this lib, instead of forcing users to install a separate extension or copy-paste code around.
see: https://discourse.charmhub.io/t/harness-recipe-capture-events/6581
for the Harness equivalent.

### CODE REVIEW

File: scenario/utils.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    """Capture all events of type `*types` (using instance checks).
+
+    If no types are passed, will capture all types.
+    """
+    def _filter(evt):
+        if not include_lifecycle_events and isinstance(evt, (CommitEvent, PreCommitEvent)):
+            return False
+        if types:
+            return isinstance(evt, types)
+        return isinstance(evt, EventBase)
```
Comment (benhoyt) [inline scenario/utils.py]: Would this ever not be true? I.e., when would an emitted event not be an instance of `EventBase`?

File: scenario/utils.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    captured = []
+    _real_emit = Framework._emit  # type: ignore # noqa # ugly
```
Comment (benhoyt) [inline scenario/utils.py:23]: I realize Python has these global capture things like `assertWarns`, but it doesn't seem great to me that this patches globally (the global `Framework._emit`). Is there some way to access the `charm.framework` instead, and only patch for that instance?

Something like:

```python
with capture_events(charm, StartEvent) as captured:
    State().trigger('start', charm)
```

Or maybe the test doesn't have access to the charm instance, only the type? In that case is there a way to work in with `trigger()` to only patch a single instance's framework._emit?

File: tests/test_e2e/test_capture_events.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    assert len(captured) == 1
+    assert isinstance(captured[0], FooEvent)
+
+
+def test_capture_lifecycle(mycharm):
+    with capture_events(FooEvent, include_lifecycle_events=True) as captured:
+        State().trigger('start', mycharm, meta=mycharm.META)
+
+    assert len(captured) == 1
+    assert isinstance(captured[0], FooEvent)
```
Comment (benhoyt) [inline tests/test_e2e/test_capture_events.py]: I think the test of `include_lifecycle_events` should actually show it's including `CommitEvent` or `PreCommitEvent`, right?

**Review by benhoyt — COMMENTED** [review]
This is interesting, and seems reasonable to me.

I've also looked at https://discourse.charmhub.io/t/harness-recipe-capture-events/6581 and it may be something we want to add to `ops.testing`. However, I want to understand the use case a bit better.

For scenario, this would really only be useful for custom events, right? Because the main event is what you `trigger()` manually, so of course that's going to be there.

For harness, this would be useful for all event types. However, I'm not sure I understand what you'd be testing? It seems we should prefer letting the event fire and the test should test that the Harness state has changed as expected. (Or with scenario, that the returned state includes whatever changes the custom event performed.)

This kind of seems like white-box testing rather than testing the charm's functionality. That can sometimes have its place, but I want to understand the motivation and use cases better.

File: tests/test_e2e/test_capture_events.py (hunk 3 above)
Comment (PietroPasotti) [reply tests/test_e2e/test_capture_events.py]: very much so

File: scenario/utils.py (hunk 1 above)
Comment (PietroPasotti) [reply scenario/utils.py]: fixed

File: scenario/utils.py (hunk 2 above)
Comment (PietroPasotti) [reply scenario/utils.py:23]: What is the danger? Framework is (in practice) a singleton, so we shouldn't need to worry about this interfering with other tests. 

As you say, at patch time there's no charm/framework instance yet, and it feels complicated to attach a hook somewhere to delay the patching to 'the right instance'.

**Comment by PietroPasotti** [comment]
> For scenario, this would really only be useful for custom events, right? 
Custom events, yes, but also 'regular' events that the charm emits of its own accord. Some charms implement the common exit hook pattern by emitting, say, update-status from `start`, `install`, etc... The captured events then would include [event being dispatched, `update-status`].

Also don't just think 'charm', but also 'charm lib'. If you're writing unittests for a charm lib (e.g. a `DatabaseProvider`), you typically want to check that the right 'custom event' is emitted by the charm lib as a response of a combination of 'the right state' and a `relation-changed` or some other event. The events being emitted by the charm are as you say not part of the output state, so this is white box testing, indeed. However it seems to be very useful to charmers, it's a highly desired assertion, so much that there was that extension lib for harness as well.

File: scenario/utils.py (hunk 2 above)
Comment (benhoyt) [reply scenario/utils.py:23]: Fair enough. If (something like) this was added to `Harness`, there is a framework instance there, so we could add it as `Harness.capture_events()` or similar. Or event just always capture events in the harness and you'd access it as `emitted_events` (you could use a list comprehension with `isinstance` to filter).

**Comment by benhoyt** [comment]
> Also don't just think 'charm', but also 'charm lib'. If you're writing unittests for a charm lib (e.g. a DatabaseProvider), you typically want to check that the right 'custom event' is emitted by the charm lib as a response of a combination of 'the right state' and a relation-changed or some other event. The events being emitted by the charm are as you say not part of the output state, so this is white box testing, indeed. However it seems to be very useful to charmers, it's a highly desired assertion, so much that there was that extension lib for harness as well.

I dunno, I prefer to have tests explicitly observe the event, so it's more like how you'd actually do it in a charm. This is straight-forward to do, as shown in the data-platform-libs tests of exactly that case. They [set up the charm like this](https://github.com/canonical/data-platform-libs/blob/704ac73c48894771171c8c2ce56dabe0d1225adb/tests/unit/test_database_provides.py#L27-L39):

```python
class DatabaseCharm(CharmBase):
    """Mock database charm to use in units tests."""

    def __init__(self, *args):
        super().__init__(*args)
        self.database = DatabaseProvides(self, RELATION_NAME)
        self.framework.observe(self.database.on.database_requested, self._on_database_requested)

    def _on_database_requested(self, _) -> None:
        pass
```

And then [use `patch()` in a test](https://github.com/canonical/data-platform-libs/blob/704ac73c48894771171c8c2ce56dabe0d1225adb/tests/unit/test_database_provides.py#L93-L110) to ensure it's called:

```python
    @patch.object(DatabaseCharm, "_on_database_requested")
    def test_on_database_requested(self, _on_database_requested):
        """Asserts that the correct hook is called when a new database is requested."""
        # Simulate the request of a new database plus extra user roles.
        self.harness.update_relation_data(
            self.rel_id,
            "application",
            {"database": DATABASE, "extra-user-roles": EXTRA_USER_ROLES},
        )

        # Assert the correct hook is called.
        _on_database_requested.assert_called_once()

        # Assert the database name and the extra user roles
        # are accessible in the providers charm library event.
        event = _on_database_requested.call_args[0][0]
        assert event.database == DATABASE
        assert event.extra_user_roles == EXTRA_USER_ROLES
```

It's maybe 2 more lines of code, but it uses builtin stdlib constructs, and exercises the actual observe + emit logic normally.

Or you could even avoid the `patch` altogether, and just have the mock charm append events to a list. I personally quite like this approach as I find it less magical than patch (I kind of find `patch` and `call_args` and all that stuff basically a custom DSL that you have to learn instead of just appending to a normal Python list :-). Like so:

```python
class DatabaseCharm(CharmBase):
    """Mock database charm to use in units tests."""

    def __init__(self, *args):
        super().__init__(*args)
        self.database = DatabaseProvides(self, RELATION_NAME)
        self.events = []
        self.framework.observe(self.database.on.database_requested, self._on_database_requested)

    def _on_database_requested(self, event) -> None:
        assert isinstance(event, DatabaseRequestedEvent)
        self.events.append(event)

# ...

    def test_database_requested_event_manual(self):
        # Test custom event creation

        # Test the event being emitted by the application.
        self.harness.update_relation_data(self.rel_id, "application", {"database": DATABASE})
        assert len(self.harness.charm.events) == 1
        event = self.harness.charm.events[0]
        assert event.app.name == "application"

        # Reset the diff data to trigger the event again later.
        self.harness.update_relation_data(self.rel_id, "database", {"data": "{}"})
        self.harness.charm.events = []

        # Test the event being emitted by the unit.
        self.harness.update_relation_data(self.rel_id, "application/0", {"database": DATABASE})
        assert len(self.harness.charm.events) == 1
        event = self.harness.charm.events[0]
        assert event.unit.name == "application/0"
```

Incidentally, that file [also uses](https://github.com/canonical/data-platform-libs/blob/704ac73c48894771171c8c2ce56dabe0d1225adb/tests/unit/test_database_provides.py#L174-L188) `capture_events.capture` -- I'm guessing tests added be two different people at different times.

**Comment by PietroPasotti** [comment]
> > Also don't just think 'charm', but also 'charm lib'. If you're writing unittests for a charm lib (e.g. a DatabaseProvider), you typically want to check that the right 'custom event' is emitted by the charm lib as a response of a combination of 'the right state' and a relation-changed or some other event. The events being emitted by the charm are as you say not part of the output state, so this is white box testing, indeed. However it seems to be very useful to charmers, it's a highly desired assertion, so much that there was that extension lib for harness as well.
> 
> I dunno, I prefer to have tests explicitly observe the event, so it's more like how you'd actually do it in a charm. This is straight-forward to do, as shown in the data-platform-libs tests of exactly that case. They [set up the charm like this](https://github.com/canonical/data-platform-libs/blob/704ac73c48894771171c8c2ce56dabe0d1225adb/tests/unit/test_database_provides.py#L27-L39):
> 
> ```python
> class DatabaseCharm(CharmBase):
>     """Mock database charm to use in units tests."""
> 
>     def __init__(self, *args):
>         super().__init__(*args)
>         self.database = DatabaseProvides(self, RELATION_NAME)
>         self.framework.observe(self.database.on.database_requested, self._on_database_requested)
> 
>     def _on_database_requested(self, _) -> None:
>         pass
> ```
> 
> And then [use `patch()` in a test](https://github.com/canonical/data-platform-libs/blob/704ac73c48894771171c8c2ce56dabe0d1225adb/tests/unit/test_database_provides.py#L93-L110) to ensure it's called:
> 
> ```python
>     @patch.object(DatabaseCharm, "_on_database_requested")
>     def test_on_database_requested(self, _on_database_requested):
>         """Asserts that the correct hook is called when a new database is requested."""
>         # Simulate the request of a new database plus extra user roles.
>         self.harness.update_relation_data(
>             self.rel_id,
>             "application",
>             {"database": DATABASE, "extra-user-roles": EXTRA_USER_ROLES},
>         )
> 
>         # Assert the correct hook is called.
>         _on_database_requested.assert_called_once()
> 
>         # Assert the database name and the extra user roles
>         # are accessible in the providers charm library event.
>         event = _on_database_requested.call_args[0][0]
>         assert event.database == DATABASE
>         assert event.extra_user_roles == EXTRA_USER_ROLES
> ```
> 
> It's maybe 2 more lines of code, but it uses builtin stdlib constructs, and exercises the actual observe + emit logic normally.
> 
> Or you could even avoid the `patch` altogether, and just have the mock charm append events to a list. I personally quite like this approach as I find it less magical than patch (I kind of find `patch` and `call_args` and all that stuff basically a custom DSL that you have to learn instead of just appending to a normal Python list :-). Like so:
> 
> ```python
> class DatabaseCharm(CharmBase):
>     """Mock database charm to use in units tests."""
> 
>     def __init__(self, *args):
>         super().__init__(*args)
>         self.database = DatabaseProvides(self, RELATION_NAME)
>         self.events = []
>         self.framework.observe(self.database.on.database_requested, self._on_database_requested)
> 
>     def _on_database_requested(self, event) -> None:
>         assert isinstance(event, DatabaseRequestedEvent)
>         self.events.append(event)
> 
> # ...
> 
>     def test_database_requested_event_manual(self):
>         # Test custom event creation
> 
>         # Test the event being emitted by the application.
>         self.harness.update_relation_data(self.rel_id, "application", {"database": DATABASE})
>         assert len(self.harness.charm.events) == 1
>         event = self.harness.charm.events[0]
>         assert event.app.name == "application"
> 
>         # Reset the diff data to trigger the event again later.
>         self.harness.update_relation_data(self.rel_id, "database", {"data": "{}"})
>         self.harness.charm.events = []
> 
>         # Test the event being emitted by the unit.
>         self.harness.update_relation_data(self.rel_id, "application/0", {"database": DATABASE})
>         assert len(self.harness.charm.events) == 1
>         event = self.harness.charm.events[0]
>         assert event.unit.name == "application/0"
> ```
> 
> Incidentally, that file [also uses](https://github.com/canonical/data-platform-libs/blob/704ac73c48894771171c8c2ce56dabe0d1225adb/tests/unit/test_database_provides.py#L174-L188) `capture_events.capture` -- I'm guessing tests added be two different people at different times.

To be fair, this PR will be redundant as soon as OF 2.0.1 releases with my 'log custom events' changes, because at that point we can tail the logs to find out what events have been fired.

**Comment by deusebio** [comment]
TLDR I personally quite like the proposal and I believe it would be useful for charmers (at least for me/us :D) . Indeed, in data platform as Pietro mentioned, we fire some custom events as part of the charmlib. Beside, depending on the pre-conditions, we might emit different events (e.g. [here](https://github.com/canonical/data-platform-libs/blob/main/lib/charms/data_platform_libs/v0/data_interfaces.py#L1083)) and the framework would provide a natural way to catch and assert logic around these important events (together with the type, metadata, etc)

Of course, we could do this in other ways, but the proposal put a nice framework around it. With this it should be more natural to also double check not only that the event was called (as in the first example), but also the metadata attached to it. I'd also prefer this over producing and parsing logs (as Pietro mentioned), as this would need people to serialize/deserialize information via strings that could naturally be available in a data structure. 

Personally, I'm not too much in favor of the second option that uses an internal, mutable `self.events` attribute. I'm not sure (apologies here for my ignorance on how harness/scenario works, so bear with me) whether that state is reset from test to test, but I can see (in general) that these state mutations can make it a bit harder to create confined envs to be tested (if not properly handled by contexts and resets), which is something I like about scenario framework in general (I mean the facts that creates a declarative way to create confined preset->action->output frame around tests)

**Comment by PietroPasotti** [comment]
>  I'd also prefer this over producing and parsing logs (as Pietro mentioned), as this would need people to serialize/deserialize information via strings that could naturally be available in a data structure.

for clarity, what I meant is that Scenario could provide a `scenario.utils.get_emitted_events(state)`
which would parse for you `state.juju_log`. So the user-facing story is not much different in terms of 'effort'.

**Comment by PietroPasotti** [comment]
archiving this one for now. There are other ways to assert event emission and they're not particularly cumbersome. We can reevaluate when there's logging for custom events, and after some usage has shown that the need for sugar is real.
